### PR TITLE
NH-2941

### PR DESCRIPTION
**Pull request by skrishna-nh**
JIRA: https://nhibernate.jira.com/browse/NH-2941
Updated the test to demonstrate that using child sessions does not help in this case.

### CODE REVIEW

File: src/NHibernate/Impl/SessionImpl.cs [hunk 1]
```diff
@@ -600,11 +600,12 @@ public override void List(string query, QueryParameters queryParameters, IList r
...
-				var plan = GetHQLQueryPlan(query, false);
-				AutoFlushIfRequired(plan.QuerySpaces);
+				var plan = GetHQLQueryPlan(query, false);
+                
+                dontFlushFromFind++; //stops flush being called multiple times if this method is recursively called
+                AutoFlushIfRequired(plan.QuerySpaces);
 
 				bool success = false;
-				dontFlushFromFind++; //stops flush being called multiple times if this method is recursively called
 				try
```
Comment (fredericDelaporte) [inline src/NHibernate/Impl/SessionImpl.cs]: The try catch statement is also about ensuring `dontFlushFromFind` get decremented before leaving the method, whatever happened. Calling `AutoFlushIfRequired(plan.QuerySpaces);` between `dontFlushFromFind` and the try is a no-go.
A dedicated try finally looks now required for ensuring `dontFlushFromFind` decrement without trigerring  `AfterOperation` if the auto flushed failed.

File: src/NHibernate/Impl/SessionImpl.cs [hunk 2]
```diff
@@ -676,10 +678,11 @@ public override IEnumerable<T> Enumerable<T>(string query, QueryParameters query
...
 				queryParameters.ValidateParameters();
-				var plan = GetHQLQueryPlan(query, true);
-				AutoFlushIfRequired(plan.QuerySpaces);
+				var plan = GetHQLQueryPlan(query, true);
+                
+                dontFlushFromFind++; //stops flush being called multiple times if this method is recursively called
+                AutoFlushIfRequired(plan.QuerySpaces);
 
-				dontFlushFromFind++; //stops flush being called multiple times if this method is recursively called
 				try
```
Comment (fredericDelaporte) [inline src/NHibernate/Impl/SessionImpl.cs]: Same flaw as in previous comment.

File: src/NHibernate/Impl/SessionImpl.cs [hunk 3]
```diff
@@ -697,10 +700,11 @@ public override IEnumerable Enumerable(string query, QueryParameters queryParame
...
```
Comment (fredericDelaporte) [inline src/NHibernate/Impl/SessionImpl.cs]: Same flaw as previous comment.

File: src/NHibernate/Impl/SessionImpl.cs [hunk 4]
```diff
@@ -2078,11 +2082,12 @@ public override void ListCustomQuery(ICustomQuery customQuery, QueryParameters q
...
-				CustomLoader loader = new CustomLoader(customQuery, Factory);
-				AutoFlushIfRequired(loader.QuerySpaces);
+				CustomLoader loader = new CustomLoader(customQuery, Factory);
+                
+                dontFlushFromFind++;
+                AutoFlushIfRequired(loader.QuerySpaces);
 
 				bool success = false;
-				dontFlushFromFind++;
 				try
```
Comment (fredericDelaporte) [inline src/NHibernate/Impl/SessionImpl.cs]: Same flaw as previous comment.

**Comment by hazzik** [comment]
This PR is 5 years old. I don't think it is going to make into master as is. So, if the feature is still desired - please implement a new PR considering the review & JIRA comments

**Pull request closed without merging**